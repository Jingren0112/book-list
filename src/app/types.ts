export interface IBook {
    book_author: string[],
    book_pages: number,
    book_publication_city: string,
    book_publication_country: string,
    book_publication_year: number,
    book_title: string,
    id: number
}

export enum loadingStatus {
    loading = 'loading',
    success = 'success',
    failed = 'failed'
}