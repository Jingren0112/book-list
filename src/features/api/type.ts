
export const url = 'http://nyx.vima.ekt.gr:3000'
export const booksEndPoint = '/api/books'