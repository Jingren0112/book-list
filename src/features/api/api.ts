import axios from 'axios'
import { booksEndPoint, param, url } from './type'

export const fetchBook = () => axios.post(url + booksEndPoint + param)
