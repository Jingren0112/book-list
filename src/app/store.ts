import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import bookListSlice from '../features/bookList/bookListSlice';

export const store = configureStore({
  reducer: {
    bookList: bookListSlice,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
