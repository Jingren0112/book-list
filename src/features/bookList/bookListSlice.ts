import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../app/store';
import { IBook, loadingStatus } from '../../app/types';
import { fetchBook } from '../api/api';

export interface IBookListState {
    data: IBook[] | null,
    status: loadingStatus,
    currentPage: number,
    itemsPerPage: number,
    error: any,
    search: string,
}

const initialState: IBookListState = {
    data: null,
    status: loadingStatus.loading,
    currentPage: 1,
    itemsPerPage: 10,
    error: null,
    search: '',
};

// Thunk function to execute the fetchBook api call and return the data
export const fetchBooksAsync = createAsyncThunk(
    'bookList/fetchBooks',
    async () => {
        const response = await fetchBook();
        // The value we return becomes the `fulfilled` action payload
        return response.data.books;
    }
);

export const bookListSlice = createSlice({
    name: 'bookList',
    initialState,
    reducers: {
        setCurrentPage: (state, action) => {
            state.currentPage = action.payload;
        },
        setPageItem: (state, action) => {
            state.itemsPerPage = action.payload;
        },
    },
    // including actions generated by createAsyncThunk or in other slices.
    extraReducers: (builder) => {
        builder
            .addCase(fetchBooksAsync.pending, (state) => {
                state.status = loadingStatus.loading;
            })
            .addCase(fetchBooksAsync.fulfilled, (state, action: PayloadAction<IBook[] | null>) => {
                state.status = loadingStatus.success
                state.data = action.payload;
            })
            .addCase(fetchBooksAsync.rejected, (state, action) => {
                state.status = loadingStatus.failed;
                state.error = action.error.message;
            });
    },
});

export const { setCurrentPage, setPageItem } = bookListSlice.actions;
export const selectData = (state: RootState) => state.bookList;

export default bookListSlice.reducer;
