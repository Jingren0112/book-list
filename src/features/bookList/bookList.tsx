import React, { useEffect, useState } from 'react';
import { useAppSelector, useAppDispatch } from '../../app/hooks';
import {
    fetchBooksAsync, selectData,
} from './bookListSlice';
import styles from './Counter.module.css';
import { isNil } from 'lodash';
import { IBook } from '../../app/types';
import { BookListItem } from '../bookListItem/bookListItem';
import { Pagination, PaginationItem } from '@mui/material';
import { Link, redirect, useLocation } from 'react-router-dom';
import { setCurrentPage } from './bookListSlice';

interface IBookListProps {
    currentPage?: number;
}

export const BookList = (props: IBookListProps): JSX.Element => {
    const state = useAppSelector(selectData);
    const location = useLocation();
    const query = new URLSearchParams(location.search);
    const page = parseInt(query.get('page') || '1', 10);
    console.log('page', page)
    const currentPage = props.currentPage ?? page;
    console.log('currentPage', currentPage)
    const start = (currentPage - 1) * state.itemsPerPage;
    const end = start + state.itemsPerPage;
    const currentData = !isNil(state.data) ? state.data?.slice(start, end) : null;
    console.log('start', start, 'end', end, 'currentData', currentData)
    const dispatch = useAppDispatch();
    return (
        <>
            <div>
                <h1>this is the home page</h1>
                <p>current page is </p>{ page }
                <div>
                    { !isNil(currentData) ? currentData.map((book, index) => {
                        return (
                            <div key={ index }>
                                <BookListItem book={ book } />
                            </div>
                        )
                    }) : <></> }
                </div >
            </div>
            <Pagination
                page={ page }
                count={ !isNil(state.data) ? state.data?.length / state.itemsPerPage! : 0 }
                renderItem={ (item) => (
                    <PaginationItem
                        component={ Link }
                        to={ `${item.page === 0 ? '' : `/bookList?page=${item.page}`}` }
                        { ...item }
                    />
                ) }
                onChange={ (e, newPage) => {
                    redirect(`/bookList?page='${newPage}'`)
                    dispatch(setCurrentPage(newPage))
                } }
            />
        </>);
}

