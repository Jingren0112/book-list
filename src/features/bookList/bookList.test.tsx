import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { booksEndPoint, url } from '../api/type'
import reducer, { IBookListState, initialState, setCurrentPage, setPageItem } from './bookListSlice'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { BookList } from './bookList'
import { store } from '../../app/store'

// use msw to intercept the network request during the test,
// and return the response 'test' after 150ms
// when receiving a get request to the `http://nyx.vima.ekt.gr:3000/api/book` endpoint
export const handlers = [
  rest.get(url + booksEndPoint, (req, res, ctx) => {
    return res(ctx.json('test'), ctx.delay(150))
  })
]

const server = setupServer(...handlers)

// Enable API mocking before tests.
beforeAll(() => server.listen())

// Reset any runtime request handlers we may add during the tests.
afterEach(() => server.resetHandlers())

// Disable API mocking after the tests are done.
afterAll(() => server.close())

test('should handle currentPage changes', () => {
  const previousState: IBookListState = { ...initialState, currentPage: 1 }

  expect(reducer(previousState, setCurrentPage(2))).toEqual(
    { ...initialState, currentPage: 2 }
  )
})

test('should handle page item number changes', () => {
  const previousState: IBookListState = { ...initialState, itemsPerPage: 5 }

  expect(reducer(previousState, setPageItem(10))).toEqual(
    { ...initialState, itemsPerPage: 10 }
  )
})


// snapshot test


// Mock router
const mockedUsedNavigate = jest.fn();
const mockedUseLocation = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockedUsedNavigate,
  useLocation: () => mockedUseLocation
}));

describe('bookList it', () => {
  it('should render correctly with loading state', () => {
    // arrange

    // act
    const tree = render(<Provider store={ store }><BookList /></Provider>)
    // assert
    expect(tree).toMatchSnapshot()
  }),
    it('shoud render correctly with specific page', () => {
      // arrange

      // act
      const tree = render(<Provider store={ store }><BookList currentPage={ 2 } /></Provider>)
      // assert
      expect(tree).toMatchSnapshot()
    })
})
