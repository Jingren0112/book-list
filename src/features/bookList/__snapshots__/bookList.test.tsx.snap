// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookList it shoud render correctly with specific page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1cjxh5n"
      >
        <div
          class="MuiTypography-root MuiTypography-body1 css-1ubnj4m-MuiTypography-root"
        >
          <h1>
            GG Test: Book List
          </h1>
          <button>
             Click here for data grid access!
          </button>
          <ul
            class="MuiList-root MuiList-padding css-168pop2-MuiList-root"
          >
            <li
              class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
              role="separator"
            />
          </ul>
        </div>
        <nav
          aria-label="pagination navigation"
          class="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
        >
          <ul
            class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateBeforeIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateNextIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1cjxh5n"
    >
      <div
        class="MuiTypography-root MuiTypography-body1 css-1ubnj4m-MuiTypography-root"
      >
        <h1>
          GG Test: Book List
        </h1>
        <button>
           Click here for data grid access!
        </button>
        <ul
          class="MuiList-root MuiList-padding css-168pop2-MuiList-root"
        >
          <li
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
            role="separator"
          />
        </ul>
      </div>
      <nav
        aria-label="pagination navigation"
        class="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
      >
        <ul
          class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateBeforeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateNextIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`bookList it should render correctly with loading state 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1cjxh5n"
      >
        <div
          class="MuiTypography-root MuiTypography-body1 css-1ubnj4m-MuiTypography-root"
        >
          <h1>
            GG Test: Book List
          </h1>
          <button>
             Click here for data grid access!
          </button>
          <ul
            class="MuiList-root MuiList-padding css-168pop2-MuiList-root"
          >
            <li
              class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
              role="separator"
            />
          </ul>
        </div>
        <nav
          aria-label="pagination navigation"
          class="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
        >
          <ul
            class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateBeforeIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateNextIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1cjxh5n"
    >
      <div
        class="MuiTypography-root MuiTypography-body1 css-1ubnj4m-MuiTypography-root"
      >
        <h1>
          GG Test: Book List
        </h1>
        <button>
           Click here for data grid access!
        </button>
        <ul
          class="MuiList-root MuiList-padding css-168pop2-MuiList-root"
        >
          <li
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
            role="separator"
          />
        </ul>
      </div>
      <nav
        aria-label="pagination navigation"
        class="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
      >
        <ul
          class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateBeforeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateNextIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
